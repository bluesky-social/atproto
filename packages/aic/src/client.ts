import { DidableKey } from '@adxp/crypto'
import axios from 'axios'
import { Document, hashAndFindDid } from './document'
import { Create, create, Operation, UpdateOperation } from './types'
import * as cbor from '@ipld/dag-cbor'
import * as uint8arrays from 'uint8arrays'
import { check, cidForData } from '@adxp/common'

export class AicClient {
  constructor(public url: string) {}

  async getDocument(did: string): Promise<Document> {
    const res = await axios.get(`${this.url}/${did}`)
    return res.data
  }

  async getOperationLog(did: string): Promise<Operation[]> {
    const res = await axios.get(`${this.url}/log/${did}`)
    return res.data.log
  }

  async createDid(
    signingKey: DidableKey,
    recoveryKey: string,
    username: string,
    service: string,
  ): Promise<string> {
    const op: Partial<Create> = {
      type: 'create',
      signingKey: signingKey.did(),
      recoveryKey,
      username,
      service,
      prev: null,
    }
    const signedOp = await this.signOperation(op, signingKey)
    if (!check.is(signedOp, create)) {
      throw new Error('Not a valid create operation')
    }
    const did = await hashAndFindDid(signedOp)
    await axios.post(`${this.url}/${did}`, signedOp)
    return did
  }

  async makeUpdate(
    did: string,
    op: Partial<UpdateOperation>,
    signingKey: DidableKey,
  ) {
    const log = await this.getOperationLog(did)
    if (log.length === 0) {
      throw new Error(`Could not make update: DID does not exist: ${did}`)
    }
    const prev = await cidForData(log[log.length - 1])
    const signedOp = this.signOperation(
      {
        ...op,
        prev: prev.toString(),
      },
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, signedOp)
  }

  async rotateSigningKey(did: string, newKey: string, signingKey: DidableKey) {
    return this.makeUpdate(
      did,
      { type: 'rotate_signing_key', key: newKey },
      signingKey,
    )
  }

  async rotateRecoveryKey(did: string, newKey: string, signingKey: DidableKey) {
    return this.makeUpdate(
      did,
      { type: 'rotate_recovery_key', key: newKey },
      signingKey,
    )
  }

  async udpateUsername(did: string, username: string, signingKey: DidableKey) {
    return this.makeUpdate(
      did,
      { type: 'update_username', username },
      signingKey,
    )
  }

  async updateService(did: string, service: string, signingKey: DidableKey) {
    return this.makeUpdate(did, { type: 'update_service', service }, signingKey)
  }
}

export default AicClient
