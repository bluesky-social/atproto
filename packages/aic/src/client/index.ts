import axios from 'axios'
import { CID } from 'multiformats/cid'
import { DidableKey } from '@adxp/crypto'
import { check, cidForData } from '@adxp/common'
import * as operations from '../lib/operations'
import * as t from '../lib/types'

export class AicClient {
  constructor(public url: string) {}

  async getDocument(did: string): Promise<t.DocumentData> {
    const res = await axios.get(`${this.url}/${did}`)
    return res.data
  }

  async getOperationLog(did: string): Promise<t.Operation[]> {
    const res = await axios.get(`${this.url}/log/${did}`)
    return res.data.log
  }

  async createDid(
    signingKey: DidableKey,
    recoveryKey: string,
    username: string,
    service: string,
  ): Promise<string> {
    const op = await operations.create(
      signingKey,
      recoveryKey,
      username,
      service,
    )
    if (!check.is(op, t.def.createOp)) {
      throw new Error('Not a valid create operation')
    }
    const did = await operations.didForCreateOp(op)
    await axios.post(`${this.url}/${did}`, op)
    return did
  }

  async getPrev(did): Promise<CID> {
    const log = await this.getOperationLog(did)
    if (log.length === 0) {
      throw new Error(`Could not make update: DID does not exist: ${did}`)
    }
    return cidForData(log[log.length - 1])
  }

  async rotateSigningKey(did: string, newKey: string, signingKey: DidableKey) {
    const prev = await this.getPrev(did)
    const op = await operations.rotateSigningKey(
      newKey,
      prev.toString(),
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, op)
  }

  async rotateRecoveryKey(did: string, newKey: string, signingKey: DidableKey) {
    const prev = await this.getPrev(did)
    const op = await operations.rotateRecoveryKey(
      newKey,
      prev.toString(),
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, op)
  }

  async updateUsername(did: string, username: string, signingKey: DidableKey) {
    const prev = await this.getPrev(did)
    const op = await operations.updateUsername(
      username,
      prev.toString(),
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, op)
  }

  async updateService(did: string, service: string, signingKey: DidableKey) {
    const prev = await this.getPrev(did)
    const op = await operations.updateService(
      service,
      prev.toString(),
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, op)
  }
}

export default AicClient
