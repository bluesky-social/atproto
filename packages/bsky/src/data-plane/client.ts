import assert from 'node:assert'
import { randomInt } from 'node:crypto'
import * as ui8 from 'uint8arrays'
import {
  Code,
  ConnectError,
  PromiseClient,
  createPromiseClient,
  makeAnyClient,
} from '@connectrpc/connect'
import { createGrpcTransport } from '@connectrpc/connect-node'
import { getDidKeyFromMultibase } from '@atproto/identity'
import { Service } from '../proto/bsky_connect'

export type DataPlaneClient = PromiseClient<typeof Service>
type BaseClient = { lib: DataPlaneClient; url: URL }
type HttpVersion = '1.1' | '2'
const MAX_RETRIES = 3

export const createDataPlaneClient = (
  baseUrls: string[],
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
) => {
  const clients = baseUrls.map((baseUrl) => createBaseClient(baseUrl, opts))
  assert(clients.length > 0, 'no clients available')
  return makeAnyClient(Service, (method) => {
    return async (...args) => {
      let tries = 0
      let error: unknown
      let remainingClients = clients
      while (tries < MAX_RETRIES) {
        const client = randomElement(remainingClients)
        assert(client, 'no clients available')
        try {
          return await client.lib[method.localName](...args)
        } catch (err) {
          if (
            err instanceof ConnectError &&
            (err.code === Code.Unavailable || err.code === Code.Aborted)
          ) {
            tries++
            error = err
            remainingClients = getRemainingClients(remainingClients, client)
          } else {
            throw err
          }
        }
      }
      assert(error)
      throw error
    }
  }) as DataPlaneClient
}

export { Code }

export const isDataplaneError = (
  err: unknown,
  code?: Code,
): err is ConnectError => {
  if (err instanceof ConnectError) {
    return !code || err.code === code
  }
  return false
}

const createBaseClient = (
  baseUrl: string,
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
): BaseClient => {
  const { httpVersion = '2', rejectUnauthorized = true } = opts
  const transport = createGrpcTransport({
    baseUrl,
    httpVersion,
    acceptCompression: [],
    nodeOptions: { rejectUnauthorized },
  })
  return {
    lib: createPromiseClient(Service, transport),
    url: new URL(baseUrl),
  }
}

const getRemainingClients = (clients: BaseClient[], lastClient: BaseClient) => {
  if (clients.length < 2) return clients // no clients to choose from
  if (lastClient.url.port) {
    // if the last client had a port, we attempt to exclude its whole host.
    const maybeRemaining = clients.filter(
      (c) => c.url.hostname !== lastClient.url.hostname,
    )
    if (maybeRemaining.length) {
      return maybeRemaining
    }
  }
  return clients.filter((c) => c !== lastClient)
}

const randomElement = <T>(arr: T[]): T | undefined => {
  if (arr.length === 0) return
  return arr[randomInt(arr.length)]
}

export const unpackIdentityServices = (servicesBytes: Uint8Array) => {
  const servicesStr = ui8.toString(servicesBytes, 'utf8')
  if (!servicesStr) return {}
  return JSON.parse(servicesStr) as UnpackedServices
}

export const unpackIdentityKeys = (keysBytes: Uint8Array) => {
  const keysStr = ui8.toString(keysBytes, 'utf8')
  if (!keysStr) return {}
  return JSON.parse(keysStr) as UnpackedKeys
}

export const getServiceEndpoint = (
  services: UnpackedServices,
  opts: { id: string; type: string },
) => {
  const endpoint =
    services[opts.id] &&
    services[opts.id].Type === opts.type &&
    validateUrl(services[opts.id].URL)
  return endpoint || undefined
}

export const getKeyAsDidKey = (keys: UnpackedKeys, opts: { id: string }) => {
  const key =
    keys[opts.id] &&
    getDidKeyFromMultibase({
      type: keys[opts.id].Type,
      publicKeyMultibase: keys[opts.id].PublicKeyMultibase,
    })
  return key || undefined
}

type UnpackedServices = Record<string, { Type: string; URL: string }>

type UnpackedKeys = Record<string, { Type: string; PublicKeyMultibase: string }>

const validateUrl = (urlStr: string): string | undefined => {
  let url
  try {
    url = new URL(urlStr)
  } catch {
    return undefined
  }
  if (!['http:', 'https:'].includes(url.protocol)) {
    return undefined
  } else if (!url.hostname) {
    return undefined
  } else {
    return urlStr
  }
}
