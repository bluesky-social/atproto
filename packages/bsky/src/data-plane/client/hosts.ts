import { Etcd3 } from 'etcd3'
import { EtcdMap } from '../../etcd'
import { dataplaneLogger as logger } from '../../logger'

/**
 * Interface for a reactive list of hosts, i.e. for use with the dataplane client.
 */
export interface HostList {
  get: () => Iterable<string>
  onUpdate(handler: HostListHandler): void
}

type HostListHandler = (hosts: Iterable<string>) => void

/**
 * Maintains a reactive HostList based on a simple setter.
 */
export class BasicHostList implements HostList {
  private hosts: Iterable<string>
  private handlers: HostListHandler[] = []

  constructor(hosts: Iterable<string>) {
    this.hosts = hosts
  }

  get() {
    return this.hosts
  }

  set(hosts: Iterable<string>) {
    this.hosts = hosts
    this.update()
  }

  private update() {
    for (const handler of this.handlers) {
      handler(this.hosts)
    }
  }

  onUpdate(handler: HostListHandler) {
    this.handlers.push(handler)
  }
}

/**
 * Maintains a reactive HostList based on etcd key values under a given key prefix.
 * When fallback is provided, ensures that this fallback is used whenever no hosts are available.
 */
export class EtcdHostList implements HostList {
  private kv: EtcdMap
  private inner = new BasicHostList(new Set())
  private fallback: Set<string>

  constructor(etcd: Etcd3, prefix: string, fallback?: string[]) {
    this.fallback = new Set(fallback)
    this.kv = new EtcdMap(etcd, prefix)
    this.update() // init fallback if necessary
    this.kv.watcher.on('connected', (res) => {
      logger.warn(
        { watcherId: this.kv.watcher.id, header: res.header },
        'EtcdHostList connected',
      )
    })
    this.kv.watcher.on('disconnected', (err) => {
      logger.warn(
        { watcherId: this.kv.watcher.id, err },
        'EtcdHostList disconnected',
      )
    })
    this.kv.watcher.on('error', (err) => {
      logger.error({ watcherId: this.kv.watcher.id, err }, 'EtcdHostList error')
    })
  }

  async connect() {
    await this.kv.connect()
    this.update()
    this.kv.onUpdate(() => this.update())
  }

  get() {
    return this.inner.get()
  }

  private update() {
    const hosts = new Set<string>()
    for (const host of this.kv.values()) {
      if (URL.canParse(host)) {
        hosts.add(host)
      }
    }
    if (hosts.size) {
      this.inner.set(hosts)
    } else if (this.fallback.size) {
      this.inner.set(this.fallback)
    }
  }

  onUpdate(handler: HostListHandler) {
    this.inner.onUpdate(handler)
  }
}
