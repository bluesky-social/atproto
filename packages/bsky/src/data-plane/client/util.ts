import { Code, ConnectError } from '@connectrpc/connect'
import * as ui8 from 'uint8arrays'
import { getDidKeyFromMultibase } from '@atproto/identity'

export const isDataplaneError = (
  err: unknown,
  code?: Code,
): err is ConnectError => {
  if (err instanceof ConnectError) {
    return !code || err.code === code
  }
  return false
}

export const unpackIdentityServices = (servicesBytes: Uint8Array) => {
  const servicesStr = ui8.toString(servicesBytes, 'utf8')
  if (!servicesStr) return {}
  return JSON.parse(servicesStr) as UnpackedServices
}

export const unpackIdentityKeys = (keysBytes: Uint8Array) => {
  const keysStr = ui8.toString(keysBytes, 'utf8')
  if (!keysStr) return {}
  return JSON.parse(keysStr) as UnpackedKeys
}

export const getServiceEndpoint = (
  services: UnpackedServices,
  opts: { id: string; type: string },
) => {
  const endpoint =
    services[opts.id] &&
    services[opts.id].Type === opts.type &&
    validateUrl(services[opts.id].URL)
  return endpoint || undefined
}

export const getKeyAsDidKey = (keys: UnpackedKeys, opts: { id: string }) => {
  const key =
    keys[opts.id] &&
    getDidKeyFromMultibase({
      type: keys[opts.id].Type,
      publicKeyMultibase: keys[opts.id].PublicKeyMultibase,
    })
  return key || undefined
}

type UnpackedServices = Record<string, { Type: string; URL: string }>

type UnpackedKeys = Record<string, { Type: string; PublicKeyMultibase: string }>

const validateUrl = (urlStr: string): string | undefined => {
  let url
  try {
    url = new URL(urlStr)
  } catch {
    return undefined
  }
  if (!['http:', 'https:'].includes(url.protocol)) {
    return undefined
  } else if (!url.hostname) {
    return undefined
  } else {
    return urlStr
  }
}
