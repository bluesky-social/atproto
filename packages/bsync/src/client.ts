import {
  Interceptor,
  PromiseClient,
  createPromiseClient,
} from '@connectrpc/connect'
import {
  ConnectTransportOptions,
  createConnectTransport,
} from '@connectrpc/connect-node'
import { Service } from './proto/bsync_connect'

export type BsyncClient = PromiseClient<typeof Service>

export const createClient = (opts: ConnectTransportOptions): BsyncClient => {
  const transport = createConnectTransport(opts)
  return createPromiseClient(Service, transport)
}

export const authWithApiKey =
  (apiKey: string): Interceptor =>
  (next) =>
  (req) => {
    req.header.set('authorization', `Bearer ${apiKey}`)
    return next(req)
  }
