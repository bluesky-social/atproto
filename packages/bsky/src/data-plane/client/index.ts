import assert from 'node:assert'
import { randomInt } from 'node:crypto'
import {
  Code,
  ConnectError,
  PromiseClient,
  createPromiseClient,
  makeAnyClient,
} from '@connectrpc/connect'
import { createGrpcTransport } from '@connectrpc/connect-node'
import { Service } from '../../proto/bsky_connect'
import { HostList } from './hosts'

export * from './hosts'
export * from './util'

export type DataPlaneClient = PromiseClient<typeof Service>
type HttpVersion = '1.1' | '2'
const MAX_RETRIES = 3

export const createDataPlaneClient = (
  hostList: HostList,
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
) => {
  const clients = new DataPlaneClients(hostList, opts)
  return makeAnyClient(Service, (method) => {
    return async (...args) => {
      let tries = 0
      let error: unknown
      let remainingClients = clients.get()
      while (tries < MAX_RETRIES) {
        const client = randomElement(remainingClients)
        assert(client, 'no clients available')
        try {
          return await client[method.localName](...args)
        } catch (err) {
          if (
            err instanceof ConnectError &&
            (err.code === Code.Unavailable || err.code === Code.Aborted)
          ) {
            tries++
            error = err
            remainingClients = getRemainingClients(remainingClients, client)
          } else {
            throw err
          }
        }
      }
      assert(error)
      throw error
    }
  }) as DataPlaneClient
}

export { Code }

/**
 * Uses a reactive HostList in order to maintain a pool of DataPlaneClients.
 * Each DataPlaneClient is cached per host so that it maintains connections
 * and other internal state when the underlying HostList is updated.
 */
class DataPlaneClients {
  private clients: DataPlaneClient[] = []
  private clientsByHost = new Map<string, DataPlaneClient>()

  constructor(
    private hostList: HostList,
    private clientOpts: {
      httpVersion?: HttpVersion
      rejectUnauthorized?: boolean
    },
  ) {
    this.refresh()
    this.hostList.onUpdate(() => this.refresh())
  }

  get(): readonly DataPlaneClient[] {
    return this.clients
  }

  private refresh() {
    this.clients = []
    for (const host of this.hostList.get()) {
      let client = this.clientsByHost.get(host)
      if (!client) {
        client = this.createClient(host)
        this.clientsByHost.set(host, client)
      }
      this.clients.push(client)
    }
  }

  private createClient(host: string) {
    return createBaseClient(host, this.clientOpts)
  }
}

const createBaseClient = (
  baseUrl: string,
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
): DataPlaneClient => {
  const { httpVersion = '2', rejectUnauthorized = true } = opts
  const transport = createGrpcTransport({
    baseUrl,
    httpVersion,
    acceptCompression: [],
    nodeOptions: { rejectUnauthorized },
  })
  return createPromiseClient(Service, transport)
}

const getRemainingClients = (
  clients: readonly DataPlaneClient[],
  lastClient: DataPlaneClient,
) => {
  if (clients.length < 2) return clients // no clients to choose from
  return clients.filter((c) => c !== lastClient)
}

const randomElement = <T>(arr: readonly T[]): T | undefined => {
  if (arr.length === 0) return
  return arr[randomInt(arr.length)]
}
