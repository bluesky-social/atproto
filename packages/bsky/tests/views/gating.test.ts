import assert from 'assert'
import AtpAgent from '@atproto/api'
import { TestNetwork } from '@atproto/dev-env'
import {
  isNotFoundPost,
  isThreadViewPost,
} from '../../src/lexicon/types/app/bsky/feed/defs'
import { SeedClient } from '../seeds/client'
import basicSeed from '../seeds/basic'

describe('views with interaction gating', () => {
  let network: TestNetwork
  let agent: AtpAgent
  let pdsAgent: AtpAgent
  let sc: SeedClient

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'bsky_views_gating',
    })
    agent = network.bsky.getClient()
    pdsAgent = network.pds.getClient()
    sc = new SeedClient(pdsAgent)
    await basicSeed(sc)
    await network.processAll()
  })

  afterAll(async () => {
    await network.close()
  })

  it('applies gate for empty rules.', async () => {
    const post = await sc.post(sc.dids.carol, 'empty rules')
    await pdsAgent.api.app.bsky.feed.gate.create(
      { repo: sc.dids.carol, rkey: post.ref.uri.rkey },
      { post: post.ref.uriStr, createdAt: iso(), allow: [] },
      sc.getHeaders(sc.dids.carol),
    )
    await sc.reply(sc.dids.alice, post.ref, post.ref, 'empty rules reply')
    await network.processAll()
    const {
      data: { thread },
    } = await agent.api.app.bsky.feed.getPostThread(
      { uri: post.ref.uriStr },
      { headers: await network.serviceHeaders(sc.dids.alice) },
    )
    assert(isThreadViewPost(thread))
    expect(thread.post.gate).toMatchSnapshot()
    expect(thread.viewer).toEqual({ canReply: false })
    expect(thread.replies?.length).toEqual(0)
  })

  it('applies gate for mention rule.', async () => {
    const post = await sc.post(
      sc.dids.carol,
      'mention rules @carol.test @dan.test',
      [
        {
          index: { byteStart: 14, byteEnd: 25 },
          features: [
            { $type: 'app.bsky.richtext.facet#mention', did: sc.dids.carol },
          ],
        },
        {
          index: { byteStart: 26, byteEnd: 35 },
          features: [
            { $type: 'app.bsky.richtext.facet#mention', did: sc.dids.dan },
          ],
        },
      ],
    )
    await pdsAgent.api.app.bsky.feed.gate.create(
      { repo: sc.dids.carol, rkey: post.ref.uri.rkey },
      {
        post: post.ref.uriStr,
        createdAt: iso(),
        allow: [{ $type: 'app.bsky.feed.gate#mentionRule' }],
      },
      sc.getHeaders(sc.dids.carol),
    )
    await sc.reply(
      sc.dids.alice,
      post.ref,
      post.ref,
      'mention rule reply disallow',
    )
    const danReply = await sc.reply(
      sc.dids.dan,
      post.ref,
      post.ref,
      'mention rule reply allow',
    )
    await network.processAll()
    const {
      data: { thread: aliceThread },
    } = await agent.api.app.bsky.feed.getPostThread(
      { uri: post.ref.uriStr },
      { headers: await network.serviceHeaders(sc.dids.alice) },
    )
    assert(isThreadViewPost(aliceThread))
    expect(aliceThread.viewer).toEqual({ canReply: false })
    const {
      data: { thread: danThread },
    } = await agent.api.app.bsky.feed.getPostThread(
      { uri: post.ref.uriStr },
      { headers: await network.serviceHeaders(sc.dids.dan) },
    )
    assert(isThreadViewPost(danThread))
    expect(danThread.post.gate).toMatchSnapshot()
    expect(danThread.viewer).toEqual({ canReply: true })
    const [reply, ...otherReplies] = danThread.replies ?? []
    assert(isThreadViewPost(reply))
    expect(otherReplies.length).toEqual(0)
    expect(reply.post.uri).toEqual(danReply.ref.uriStr)
  })

  it('applies gate for following rule.', async () => {
    const post = await sc.post(sc.dids.carol, 'following rule')
    await pdsAgent.api.app.bsky.feed.gate.create(
      { repo: sc.dids.carol, rkey: post.ref.uri.rkey },
      {
        post: post.ref.uriStr,
        createdAt: iso(),
        allow: [{ $type: 'app.bsky.feed.gate#followingRule' }],
      },
      sc.getHeaders(sc.dids.carol),
    )
    // carol only follows alice
    await sc.reply(
      sc.dids.dan,
      post.ref,
      post.ref,
      'following rule reply disallow',
    )
    const aliceReply = await sc.reply(
      sc.dids.alice,
      post.ref,
      post.ref,
      'following rule reply allow',
    )
    await network.processAll()
    const {
      data: { thread: danThread },
    } = await agent.api.app.bsky.feed.getPostThread(
      { uri: post.ref.uriStr },
      { headers: await network.serviceHeaders(sc.dids.dan) },
    )
    assert(isThreadViewPost(danThread))
    expect(danThread.viewer).toEqual({ canReply: false })
    const {
      data: { thread: aliceThread },
    } = await agent.api.app.bsky.feed.getPostThread(
      { uri: post.ref.uriStr },
      { headers: await network.serviceHeaders(sc.dids.alice) },
    )
    assert(isThreadViewPost(aliceThread))
    expect(aliceThread.post.gate).toMatchSnapshot()
    expect(aliceThread.viewer).toEqual({ canReply: true })
    const [reply, ...otherReplies] = aliceThread.replies ?? []
    assert(isThreadViewPost(reply))
    expect(otherReplies.length).toEqual(0)
    expect(reply.post.uri).toEqual(aliceReply.ref.uriStr)
  })

  it('applies gate for list rule.', async () => {})

  it('applies gate for multiple rules.', async () => {
    const post = await sc.post(sc.dids.carol, 'multi rules @dan.test', [
      {
        index: { byteStart: 12, byteEnd: 21 },
        features: [
          { $type: 'app.bsky.richtext.facet#mention', did: sc.dids.dan },
        ],
      },
    ])
    await pdsAgent.api.app.bsky.feed.gate.create(
      { repo: sc.dids.carol, rkey: post.ref.uri.rkey },
      {
        post: post.ref.uriStr,
        createdAt: iso(),
        allow: [
          { $type: 'app.bsky.feed.gate#mentionRule' },
          { $type: 'app.bsky.feed.gate#followingRule' },
        ],
      },
      sc.getHeaders(sc.dids.carol),
    )
    // carol only follows alice, and the post mentions dan.
    await sc.reply(sc.dids.bob, post.ref, post.ref, 'multi rule reply disallow')
    const aliceReply = await sc.reply(
      sc.dids.alice,
      post.ref,
      post.ref,
      'multi rule reply allow (following)',
    )
    const danReply = await sc.reply(
      sc.dids.dan,
      post.ref,
      post.ref,
      'multi rule reply allow (mention)',
    )
    await network.processAll()
    const {
      data: { thread: bobThread },
    } = await agent.api.app.bsky.feed.getPostThread(
      { uri: post.ref.uriStr },
      { headers: await network.serviceHeaders(sc.dids.bob) },
    )
    assert(isThreadViewPost(bobThread))
    expect(bobThread.viewer).toEqual({ canReply: false })
    const {
      data: { thread: aliceThread },
    } = await agent.api.app.bsky.feed.getPostThread(
      { uri: post.ref.uriStr },
      { headers: await network.serviceHeaders(sc.dids.alice) },
    )
    assert(isThreadViewPost(aliceThread))
    expect(aliceThread.viewer).toEqual({ canReply: true })
    const {
      data: { thread: danThread },
    } = await agent.api.app.bsky.feed.getPostThread(
      { uri: post.ref.uriStr },
      { headers: await network.serviceHeaders(sc.dids.dan) },
    )
    assert(isThreadViewPost(danThread))
    expect(danThread.post.gate).toMatchSnapshot()
    expect(danThread.viewer).toEqual({ canReply: true })
    const [reply1, reply2, ...otherReplies] = aliceThread.replies ?? []
    assert(isThreadViewPost(reply1))
    assert(isThreadViewPost(reply2))
    expect(otherReplies.length).toEqual(0)
    expect(reply1.post.uri).toEqual(danReply.ref.uriStr)
    expect(reply2.post.uri).toEqual(aliceReply.ref.uriStr)
  })

  it('applies gate for missing rules, takes no action.', async () => {
    const post = await sc.post(sc.dids.carol, 'missing rules')
    await pdsAgent.api.app.bsky.feed.gate.create(
      { repo: sc.dids.carol, rkey: post.ref.uri.rkey },
      { post: post.ref.uriStr, createdAt: iso() },
      sc.getHeaders(sc.dids.carol),
    )
    const aliceReply = await sc.reply(
      sc.dids.alice,
      post.ref,
      post.ref,
      'missing rules reply',
    )
    await network.processAll()
    const {
      data: { thread },
    } = await agent.api.app.bsky.feed.getPostThread(
      { uri: post.ref.uriStr },
      { headers: await network.serviceHeaders(sc.dids.alice) },
    )
    assert(isThreadViewPost(thread))
    expect(thread.post.gate).toMatchSnapshot()
    expect(thread.viewer).toEqual({ canReply: true })
    const [reply, ...otherReplies] = thread.replies ?? []
    assert(isThreadViewPost(reply))
    expect(otherReplies.length).toEqual(0)
    expect(reply.post.uri).toEqual(aliceReply.ref.uriStr)
  })

  it('applies gate after root post is deleted.', async () => {
    // @NOTE also covers rule application more than one level deep
    const post = await sc.post(sc.dids.carol, 'following rule w/ post deletion')
    await pdsAgent.api.app.bsky.feed.gate.create(
      { repo: sc.dids.carol, rkey: post.ref.uri.rkey },
      {
        post: post.ref.uriStr,
        createdAt: iso(),
        allow: [{ $type: 'app.bsky.feed.gate#followingRule' }],
      },
      sc.getHeaders(sc.dids.carol),
    )
    // carol only follows alice
    const orphanedReply = await sc.reply(
      sc.dids.alice,
      post.ref,
      post.ref,
      'following rule reply allow',
    )
    await pdsAgent.api.app.bsky.feed.post.delete(
      { repo: sc.dids.carol, rkey: post.ref.uri.rkey },
      sc.getHeaders(sc.dids.carol),
    )
    await network.processAll()
    await sc.reply(
      sc.dids.dan,
      post.ref,
      orphanedReply.ref,
      'following rule reply disallow',
    )
    const aliceReply = await sc.reply(
      sc.dids.alice,
      post.ref,
      orphanedReply.ref,
      'following rule reply allow',
    )
    await network.processAll()
    const {
      data: { thread: danThread },
    } = await agent.api.app.bsky.feed.getPostThread(
      { uri: orphanedReply.ref.uriStr },
      { headers: await network.serviceHeaders(sc.dids.dan) },
    )
    assert(isThreadViewPost(danThread))
    expect(danThread.viewer).toEqual({ canReply: false })
    const {
      data: { thread: aliceThread },
    } = await agent.api.app.bsky.feed.getPostThread(
      { uri: orphanedReply.ref.uriStr },
      { headers: await network.serviceHeaders(sc.dids.alice) },
    )
    assert(isThreadViewPost(aliceThread))
    assert(
      isNotFoundPost(aliceThread.parent) &&
        aliceThread.parent.uri === post.ref.uriStr,
    )
    expect(aliceThread.post.gate).toMatchSnapshot()
    expect(aliceThread.viewer).toEqual({ canReply: true })
    const [reply, ...otherReplies] = aliceThread.replies ?? []
    assert(isThreadViewPost(reply))
    expect(otherReplies.length).toEqual(0)
    expect(reply.post.uri).toEqual(aliceReply.ref.uriStr)
  })

  it('does not apply gate to original poster.', async () => {})

  it('displays gated posts not in the context of a reply.', async () => {
    // in author feed
    // in post thread at depth 0
  })

  // @TODO check inner post uri value
  it('does not apply gate unless it matches post rkey.', async () => {})
})

const iso = (date = new Date()) => date.toISOString()
