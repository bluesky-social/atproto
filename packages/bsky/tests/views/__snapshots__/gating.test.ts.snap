// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views with interaction gating applies gate after root post is deleted. 1`] = `undefined`;

exports[`views with interaction gating applies gate for empty rules. 1`] = `
Object {
  "allow": Array [],
}
`;

exports[`views with interaction gating applies gate for following rule. 1`] = `
Object {
  "allow": Array [
    Object {
      "$type": "app.bsky.feed.defs#followingRuleView",
    },
  ],
}
`;

exports[`views with interaction gating applies gate for mention rule. 1`] = `
Object {
  "allow": Array [
    Object {
      "$type": "app.bsky.feed.defs#mentionRuleView",
    },
  ],
}
`;

exports[`views with interaction gating applies gate for missing rules, takes no action. 1`] = `Object {}`;

exports[`views with interaction gating applies gate for multiple rules. 1`] = `
Object {
  "allow": Array [
    Object {
      "$type": "app.bsky.feed.defs#mentionRuleView",
    },
    Object {
      "$type": "app.bsky.feed.defs#followingRuleView",
    },
  ],
}
`;
